### PR TITLE
Feature/129158777 enable international sales order

### PR DESCRIPTION
**Pull request by j7nguyen**
## Description

Updates to gem that allow it to process international sales orders. Add comments to code and refactor to reduce redundant code and appease Rubocop.
## Changes
- [ ] update `SalesOrder`, `SalesOrderHeader`, `SalesOrderShipmentInformation`, `SalesOrderDetail`, and `SalesOrderLineItem` to accommodate additional xml fields needed for international shipments
- [ ] add `SalesOrderLineItemNameValue`, `SalesOrderHeaderNameValue`, and `SalesOrderLineItemSpecialMessage` as new elements in a sales order
- [ ] remove `customer_id` from `BaseElement` since it's not needed by all elements.
- [ ] replace sales order XSD file with recently provided file by Ingram Micro
- [ ] refactor other classes to clean up code and reduce redundancy

[#129158777]

### CODE REVIEW

File: lib/ingram_micro/elements/credit_card_information.rb [hunk 1]
```diff
@@ -17,7 +17,7 @@ class IngramMicro::CreditCardInformation < IngramMicro::BaseElement
     billing_zip_match: nil,
     avs_hold: nil,
     merchant_name: nil
-  }
+  }.freeze
```
Comment (sent-hil) [inline lib/ingram_micro/elements/credit_card_information.rb:20]: 👍

File: lib/ingram_micro/elements/outbound_purchase_order_information.rb [hunk 2]
```diff
@@ -6,14 +6,7 @@ class IngramMicro::OutboundPurchaseOrderInformation < IngramMicro::BaseElement
     purchase_order_event: nil,
     currency_code: nil,
     comments: nil
-  }
-
-  def initialize(options = {})
```
Comment (sent-hil) [inline lib/ingram_micro/elements/outbound_purchase_order_information.rb:11]: why was this removed? irc i added this since we didn't need a `customer_id` here and having it there was problematic

File: spec/ingram_micro/elements/sales_order_line_item_special_message_spec.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
+require 'spec_helper'
+require 'pry'
+
+describe IngramMicro::SalesOrderLineItemSpecialMessage do
+
```
Comment (sent-hil) [inline spec/ingram_micro/elements/sales_order_line_item_special_message_spec.rb]: extra newline

File: lib/ingram_micro/elements/return_authorization_submission.rb [hunk 4]
```diff
@@ -1,67 +1,69 @@
...
-    builder.send('detail') do
-      @element[:detail].build(builder)
     end
-  end
 
-  def check_line_items
-    if @element[:detail].element[:line_items].empty?
-      line_item = IngramMicro::ReturnAuthorizationLineItem.new
-      @element[:detail].element[:line_items] << line_item
+    def check_line_items
```
Comment (sent-hil) [inline lib/ingram_micro/elements/return_authorization_submission.rb]: method does more than just `check`, should be renamed to convey that it does more

File: lib/ingram_micro/elements/sales_order_line_item.rb [hunk 5]
```diff
@@ -1,37 +1,79 @@
...
+        next if field == (:special_message || :line_name_value)
+        element_name = field.to_s.tr('_', '-')
+        element_value = formatted_value_of(field)
+        builder.send(element_name, element_value)
+      end
+      add_special_message(builder)
+      # Expect that line_name_value will be an array of arrays, each of which
+      # contains a name and value that will translate to line-attribute-name and
+      # line-attribute-value.
+      unless element[:line_name_value].nil?
```
Comment (sent-hil) [inline lib/ingram_micro/elements/sales_order_line_item.rb]: couldn't this be written as `if element[:line_name_value]`

File: lib/ingram_micro/elements/sales_order_line_item.rb [hunk 6]
```diff
@@ -1,37 +1,79 @@
...
+        element[:line_name_value].each do |pair|
+          attr_name, attr_value = pair
```
Comment (sent-hil) [inline lib/ingram_micro/elements/sales_order_line_item.rb:57]: did you do this on purpose? `attr_value` is assigned to nil

File: lib/ingram_micro/elements/sales_order_line_item.rb [hunk 7]
```diff
@@ -1,37 +1,79 @@
...
 
-  def defaults
-    DEFAULTS
+    # The <special-message> element is optional, including some information
+    # pertaining to international shipments. An instance of the class
+    # IngramMicro::SalesOrderLineItemSpecialMessage is passed in on
+    # initialization to make this work.
+    def add_special_message(builder)
+      message = element[:special_message]
+      message.build(builder) unless message.nil?
```
Comment (sent-hil) [inline lib/ingram_micro/elements/sales_order_line_item.rb]: `if message`

File: lib/ingram_micro/base_element.rb [hunk 8]
```diff
@@ -1,18 +1,22 @@
...
 
-  def initialize(options={})
+  def initialize(options = {})
     @element = {}
-    self.defaults.each do |field, default|
+    defaults.each do |field, default|
       value = (options[field] ? options[field] : default)
       @element[field] = value
     end
-    @element[:customer_id] ||= IngramMicro.configuration.customer_id
```
Comment (sent-hil) [inline lib/ingram_micro/base_element.rb:10]: wouldn't removing this be problematic? are you sure this doesn't break the xml we send to IM

File: lib/ingram_micro/elements/return_authorization_submission.rb [hunk 9]
```diff
@@ -1,67 +1,69 @@
...
-    purchase_order_information: nil,
-    customer_id: nil,
-    business_name: nil,
-    carrier_name: nil
-  }
-
-  def defaults
-    DEFAULTS
-  end
+    def defaults
```
Comment (sent-hil) [inline lib/ingram_micro/elements/return_authorization_submission.rb:15]: is there a reason you're redefining this in each class? if you only put it in base class and sub class it, does it look up the right `DEFAULT` ?

File: lib/ingram_micro/elements/sales_order_line_item_name_value.rb [hunk 10]
```diff
@@ -0,0 +1,33 @@
+module IngramMicro
+  class SalesOrderLineItemNameValue < BaseElement
+    ATTRIBUTE_NAMES = [
```
Comment (sent-hil) [inline lib/ingram_micro/elements/sales_order_line_item_name_value.rb]: what's the difference between this and `HEADER_ATTRIBUTE_NAMES`

File: lib/ingram_micro/elements/sales_order_submission.rb [hunk 11]
```diff
@@ -1,67 +1,68 @@
...
-    builder.send('detail') do
-      @element[:detail].build(builder)
-    end
-  end
 
-  def check_line_items
-    if @element[:line_items].empty?
-      line_item = IngramMicro::SalesOrderLineItem.new
-      @element[:line_items] << line_item
+    def check_line_items
```
Comment (sent-hil) [inline lib/ingram_micro/elements/sales_order_submission.rb]: method does more than just check, should be renamed to convey that it does more

File: lib/ingram_micro/elements/shipment_status.rb [hunk 12]
```diff
@@ -1,44 +1,47 @@
...
-    builder.send('detail') do
-      @element[:detail].build(builder)
-    end
-  end
 
-  def check_line_items
-    if @element[:line_items].empty?
-      ss_line_item = IngramMicro::ShipmentStatusLineItem.new
-      @element[:line_items] << ss_line_item
+    def check_line_items
```
Comment (sent-hil) [inline lib/ingram_micro/elements/shipment_status.rb]: method does more than just check, should be renamed to convey that it does more

File: lib/ingram_micro/elements/shipment_status.rb [hunk 13]
```diff
@@ -1,44 +1,47 @@
...
+        @element[:line_items] << ss_line_item
+      end
     end
-  end
 
-  def valid?
-      raise IngramMicro::InvalidType.new('customer_id must be a number') unless integer?(@element[:customer_id])
-    true
+    def valid?
+      unless integer?(@element[:customer_id])
```
Comment (sent-hil) [inline lib/ingram_micro/elements/shipment_status.rb:33]: feels weird having a private method called `integer?` does customer_id have to be a Integer but in a string?

File: lib/ingram_micro/transmission.rb [hunk 14]
```diff
@@ -8,19 +8,24 @@ class IngramMicro::Transmission
...
 
-  def initialize(options={})
+  def initialize(options = {})
     @errors = []
   end
 
+  # schema_valid? compares an xml document (the result of self.xml_builder) to
+  # the appropriate xsd (XML Schema Definition).
+  # NOTE: the order of the fields matters when checking against the xsd files.
+  #   So, if in the order-header of a sales order, if customer_channel_type
```
Comment (sent-hil) [inline lib/ingram_micro/transmission.rb]: include `, but if customer_channel_type appears at end of xsd file, it will fail...`

File: lib/ingram_micro/transmission.rb [hunk 15]
```diff
@@ -44,12 +49,12 @@ def add_transaction_info(builder)
   end
 
   def submit_request
-    raise IngramMicro::XMLSchemaMismatch, 'xml did not pass schema' if !schema_valid?
+    raise IngramMicro::XMLSchemaMismatch, 'xml did not pass schema' unless schema_valid?
     send_request
   end
 
   def send_request
     client = IngramMicro::Client.new
```
Comment (sent-hil) [inline lib/ingram_micro/transmission.rb]: not need to store this in a new variable, can be inlined

File: spec/ingram_micro/elements/sales_order_line_item_spec.rb [hunk 16]
```diff
@@ -0,0 +1,98 @@
+require 'spec_helper'
+require 'pry'
+
+describe IngramMicro::SalesOrderLineItem do
+  # let(:populated_sales_order) { Fabricate.build(:sales_order) }
```
Comment (sent-hil) [inline spec/ingram_micro/elements/sales_order_line_item_spec.rb]: remove this and line below

File: spec/ingram_micro/elements/sales_order_line_item_spec.rb [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+  # let(:line_item_1) { populated_sales_order.detail.element[:line_items].first }
+  let(:line_item) { Fabricate.build(:sales_order_line_item) }
+  describe '#line_no and #line_no=' do
```
Comment (sent-hil) [inline spec/ingram_micro/elements/sales_order_line_item_spec.rb:7]: space between let and describe

File: spec/ingram_micro/elements/sales_order_line_item_spec.rb [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+    it "Sets the line_no" do
+      line_item.element[:line_no] = 4
+
+      expect(line_item.line_no).to eq 4
+    end
+  end
+
+  describe '#add_line_name_value' do
+    let(:line_attr_name) { 'international-declared-value' }
+    let(:line_attr_value) { '1999' }
```
Comment (sent-hil) [inline spec/ingram_micro/elements/sales_order_line_item_spec.rb:21]: space between let and context

File: spec/ingram_micro/elements/sales_order_line_item_spec.rb [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+        special_message1: 'duuuude'
+        })
+      li = described_class.new(special_message: message)
+      builder.send('message') do
+        li.build(builder)
+      end
+      expect(builder.to_xml).to include('<special-message>')
+      expect(builder.to_xml).to include('<engraving-font>hellavetica')
+    end
+
```
Comment (sent-hil) [inline spec/ingram_micro/elements/sales_order_line_item_spec.rb]: unnecessary new line

File: spec/ingram_micro/elements/sales_order_line_item_spec.rb [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+          li = described_class.new(line_name_value: [["international-license-value",'300.50']])
+          builder = Nokogiri::XML::Builder.new
+          expect(li).to receive(:add_line_name_value)
+          builder.send('message') do
+            li.build(builder)
+          end
+        end
+
+        it 'properly creates xml' do
+          li = described_class.new(product_name: "killerizer", line_name_value: [["international-license-value",'300.50']])
```
Comment (sent-hil) [inline spec/ingram_micro/elements/sales_order_line_item_spec.rb:61]: i personally don't like described_class because each time i've to scroll up and see what the class is...not need to change it, just a fyi

**Comment by sent-hil** [comment]
@vietjew done reviewing. If any of the comments don't make sense, we can pair tuesday. Please assign it back to me when done fixing/replying to the comments.

File: lib/ingram_micro/transmission.rb (hunk 15 above)
Comment (j7nguyen) [reply lib/ingram_micro/transmission.rb]: Good call, fixed.

File: spec/ingram_micro/elements/sales_order_line_item_spec.rb (hunk 20 above)
Comment (j7nguyen) [reply spec/ingram_micro/elements/sales_order_line_item_spec.rb:61]: I see what you mean, but with these really long class names I find described_class helpful.

File: lib/ingram_micro/elements/shipment_status.rb (hunk 12 above)
Comment (j7nguyen) [reply lib/ingram_micro/elements/shipment_status.rb]: I had just rewritten this to appease rubocop, but now that I look more closely I don't think this method offers desired behavior. It's adding an empty line item if there are none, whereas it should raise an error.

File: lib/ingram_micro/elements/sales_order_submission.rb (hunk 11 above)
Comment (j7nguyen) [reply lib/ingram_micro/elements/sales_order_submission.rb]: Removing this altogether so that if no line items are passed in, the xml won't pass the schema.

File: lib/ingram_micro/elements/sales_order_line_item_name_value.rb (hunk 10 above)
Comment (j7nguyen) [reply lib/ingram_micro/elements/sales_order_line_item_name_value.rb]: They're the acceptable values for a different element, which perhaps should be implied more in the constant name.

File: lib/ingram_micro/base_element.rb (hunk 8 above)
Comment (j7nguyen) [reply lib/ingram_micro/base_element.rb:10]: I think this is better, as we don't need `customer_id` set on every element. This way, we are more explicit about setting it in the necessary places.

File: lib/ingram_micro/elements/sales_order_line_item.rb (hunk 6 above)
Comment (j7nguyen) [reply lib/ingram_micro/elements/sales_order_line_item.rb:57]: It's being assigned to the second element in pair, which is an array. Perhaps this isn't obvious and I should instead do `attr_name = pair[0]` and `attr_value = pair[1]`

File: lib/ingram_micro/elements/return_authorization_submission.rb (hunk 4 above)
Comment (j7nguyen) [reply lib/ingram_micro/elements/return_authorization_submission.rb]: Removing this entirely as it is undesired behavior.

File: lib/ingram_micro/elements/outbound_purchase_order_information.rb (hunk 2 above)
Comment (j7nguyen) [reply lib/ingram_micro/elements/outbound_purchase_order_information.rb:11]: I changed BaseElement, from which this inherits, to not automatically assign a `customer_id` to each element.

File: lib/ingram_micro/elements/return_authorization_submission.rb (hunk 9 above)
Comment (sent-hil) [reply lib/ingram_micro/elements/return_authorization_submission.rb:15]: @vietjew missed ^ comment
